### DISCUSSION TITLE
Federation Implementation

### DISCUSSION THREAD

**Discussion by calvinrp** [bytecodealliance/registry#215] [General]
Federation is one of the few outstanding protocol features not implemented.

Effort is underway to update the docs with targeted use cases. To highlight two use cases:

- Registry packages that import packages published to other registries;
- Registry mirrors;

This proposed implementation can be thought of as a composition of three new features:

- Proxying on behalf of another registry;
- Mapping namespace to published registry;
- Aliasing registry namespaces;

### Registry Proxying
Simply adds a request header (perhaps named `Warg-Registry`) that specifies the registry domain name that is the *registry subject* of the API request. If that header is present, the registry server would either respond like a read-thru cache of the other registry's API or error if that registry is not supported. A registry could choose to support (or not) passing along publish requests to the *subject registry*.

Since this is just an HTTP request header, there are many implementation options. For instance, a proxy service could sit in front of the registry servers and route the request based upon the header.

Registries are expected to be selective which other registries (if any) that they support proxying.

### Namespace Mapping
A package identifier, such as `my-company:foo-package`, does not indicate the registry domain name where it is published. Also, namespaces (`my-company` in this example) are not expected to be globally unique across all registries.

The client CLI needs to know which registry has which package. Also, when publishing packages, a registry needs to validate package imports and know which registry has which package.

This implementation proposes using namespaces to group package identifiers and associate them with their published registry domain name.

This proposes a new registry API endpoint for the client to retrieve the registry's namespace mappings to its published registry domain names. This is the mappings that the registry will use to validate package imports published to the registry. The registry will also indicate if it supports proxying those registry domain names.

The client CLI can use the mappings provided by the registry as well as extend them or override them. This configuration can be set for user profile global use or project-based with a lock file. The lock file includes the checkpoints for each registry.

### Namespace Aliasing
The following proposes an extension to the namespace mapping configuration.

To resolve naming conflicts, namespace mappings can alias the source registry's namespace to a new name to be used by packages published to its registry.

By default, package imports are resolved in the context of the mappings of the registry it was published. To resolve package imports as if it was published to a different registry, the namespace mappings can specify a registry context.

**Comment by lann** [comment]
> perhaps named `X-Registry`

The `x-` prefix is [no longer recommended](https://datatracker.ietf.org/doc/html/rfc6648). Maybe `Warg-Registry`?

**Reply by calvinrp** [reply]
Right. Old habits die hard. Works for me.

**Comment by lann** [comment]
I'm still very fuzzy on how namespace mapping/aliasing would work in practice.

If I publish a package with a dependency on `some-namspaced:package` with the intention that it points at a specific implementation, who can override that intention? It makes sense that the end-developer would be able to, but what about a registry?

I can imagine scenarios where the registry I published to would be able to replace the dependency (via a "trust reset" of the package name); I have already trusted the registry to some extent to manage _my_ package name so that doesn't seem like a major leap.

Allowing a proxying registry to override a dependency is harder to reason about; certainly there are use-cases for it (my company needs to swap in a dependency hotfix) but I'm not sure how to balance that against a potential for misuse.

This may ultimately be more of a question of UX than technical features, but I think in this kind of design spec'ing the UX can be as important as protocol design.

**Reply by calvinrp** [reply]
I share your concerns. But it may be OK. A few thoughts...

If the registry is controlled by you, your team/project or your company, that level of control at the registry level may be desired.

If you want different registry behavior, you can achieve that by configuring your own registry and setting namespace mappings. Assuming we make it easy and cheap to do so.

Also, clients can override namespace mappings in a lock file that is committed to a project repo.

**Reply by lann** [reply]
Trying to get a bit more concrete, let's think about a simple (hypothetical) client interaction: `warg show-deps <package-spec> --recursive`. Assuming client-side resolution, the client would:

1. resolve `<package-spec>` to some registry using local configuration; we'll call this the package's "origin registry"
1. fetch the package log from the origin registry (or use cached state)
1. resolve `<package-spec>` to a release using the package log
1. (print the resolved version)
1. fetch and parse the release artifact (or use previously-cached metadata)
1. for each package dependency:
   1. check for local "dependency registry override" configuration
      - if there is an override, set it as the dependency's "origin registry"
      - if there is not an override, inherit the parent package's "origin registry"
   1. recurse to step 2

Importantly, the whole process is deterministic for a particular local config and registry checkpoint(s)/cached state.

I think we can get a lot of flexibility from the details of the "dependency registry override" step:

- a developer could explicitly override a dependency globally or per-project
- an organization could (out of band) push global overrides to developers
- a registry could publish override "proposals", which a client would prompt the end-developer to copy into local config

**Reply by calvinrp** [reply]
Ooh, I really like the 3rd bullet for the CLI flow and messaging. 

I'd add to `2. fetch the package log from the origin registry (or use cached state)`, that it could be proxied.

Otherwise, that matches my understanding of the proposed implementation.

**Reply by lann** [reply]
Would "registry override proposals" be able to serve the requirements of "Namespace Aliasing"?

**Reply by calvinrp** [reply]
I think it is all the same thing. The "registry override proposals" would be a change to the registry's Namespace Mapping.

A simple namespace mapping would be mapping namespace (without renaming) to the registry domain name where its packages are published. An aliased namespace mapping would rename the namespace and optionally specify that the registry context should be the current registry instead of the source registry where the packages were originally published.

"Namespace Aliasing" is just an extension of "Namespace Mapping" with additional options.

**Reply by lukewagner** [reply]
Thinking about how overrides could work a bit more, I think we could get away with a design that doesn't involve any "override proposals" that flow downward but still enables the basic override use cases.  So the idea is that when registry `a.com` (defining namespace `a`) imports registry `b.com` (defining namespace `b`), in the basic case, this means that, for packages published to `a.com` and for clients using `a.com` as their primary registry, all of `b.com`'s packages are available with their `b.com`-defined names and contents in the `b` namespace.  However, *as part of* `a.com`'s import of `b.com`, `a.com` can attach any number of *overrides* which modify how `b.com`'s contents appear in the `b` namespace in the context of `a.com`.  By way of metaphor to, say, JS modules, where `a.com` is the following JS module, an override is like `a.com` doing:
```js
import b from "b.com";
b = { ...b, foo: overridenFoo };
```
and all the packages and clients of `a.com` now only see the overriden `b` where the package `foo` has the `overridenFoo` contents.

Lastly, a key part of how this would have to work is that, when a registry client with primary registry `a.com` does a depsolve, the depsolve algorithm does each name lookup starting at the primary registry and picking up any overrides on the path from the registry to the package containing the dependency.  Now, because namespace imports form DAGs, which mean there can be diamonds, there could be two paths from the primary registry to a particular package, and each path could have a different, conflicting override for the same name.  I think this case can just yield an error that is fixable by the primary registry "picking a winner" by installing its own override (which would unambiguously dominate them both).

So two consequences:
* In the registry's data model, I think overrides become nested under namespace imports which then creates a symmetry with how packages are nested under a locally-defined namespace.  And maybe this suggests that each override gets something like a package log (with published events, etc) which makes sense when you start thinking of serious large-scale uses of this feature.
* I don't see a situation in which a registry needs to propose an override to clients; clients just sortof inherit the overrides via the above depsolve name resolution logic.

WDYT?

**Reply by calvinrp** [reply]
Interesting. I need to digest and think through. Do you see overrides working on namespaces as well as for specific package identifiers?

**Reply by lukewagner** [reply]
In my framing above, there are two distinct things you can do: "import" a namespace (from one registry into another registry), and "override" (a package in an imported namespace).  Given this framing, I don't see where the need to "override" a namespace would arise, but I think perhaps it addresses the use cases you're imagining?

Also, just to paint a richer picture: I could imagine there being multiple distinct override *operations* (such that an "override" was  like an instruction that selected one of these operations and supplied immediate operands): overriding a specific version (effectively just swapping out the contents); overriding a whole package (replacing the whole package log); hiding a specific package (as if it wasn't in the namespace); hiding *all* packages that are not on a given list.

**Reply by esoterra** [reply]
These overrides you're describing are a bit different from the aliases I was proposing and that may be part of why we disagreed on terminology yesterday. I use the term alias to describe the operation "import namespace \<A\> as \<B\>", which is useful for example if you want to include packages from multiple other registries but they define overlapping namespaces.

I've so far been hoping to avoid getting too deep into per-package overrides/mappings (until we're a little further in) because the volume of information required to represent this will be quite large, embedding it efficiently in the data structures will be complicated, and not embedding it in the data structures opens up some security concerns.

**Reply by calvinrp** [reply]
Overrides, as described here, feel like they could be added as a feature later on. Or do you think it needs to be tackled at the start?

Conceptually, something like this might get us most of the way there:
```javascript
import { "b-namespace" as "name-for-b-in-this-registry" } from "registry.b.com"
import { "c-namespace" } from "registry.c.com"
```

**Reply by lukewagner** [reply]
@esoterra Ah hah, that helps explain alot.  Yes, "alias" in your original meaning makes more sense and is quite distinct from "overrides".  I think "import" is the closest match to module terminology (as all 3 of our metaphors show), so maybe we could all use that term going forward?  (In theory, we could add an "alias" feature to registries in the future that was exactly like component-model aliases: it would give you the ability to take an existing package and re-publish it (its content and log) with a second name (in the same or different namespace), which you could imagine being useful in certain esoteric situations.)

@calvinrp I agree that overrides (at least as I described them above) could be added later on; I was mostly just sketching them out to help build a shared understanding of what this "override" is and how it relates to a near-er term "import" feature that seems like the foundation of federation.

**Reply by calvinrp** [reply]
"Import namespace", naming convention works for me. @lukewagner Thanks for clarifying and thinking through.

**Comment by calvinrp** [comment]
I was proposing Namespace Mapping and Aliasing. Do we need finer grain mapping/aliasing down to the Package Identifier?

**Reply by lann** [reply]
If it is mostly a client-side concern I think it can be "whatever the client wants to do". Certainly as an end developer I might want to remap one specific package to a fork.

**Reply by calvinrp** [reply]
Yeah, I was thinking the same. Assume we would want a registry's mappings to have that control as well.

**Reply by esoterra** [reply]
I think I agree with Lann in the sense that a client can choose where/how to lookup packages however they want including at the individual package level. I don't however agree with Calvin in that I think it's much much more practical/useful/obvious for registries to map things at the namespace level and produce the general property that things from the namespace come from the same origin.

i.e. I think the protocol should concern itself only with defining mappings from namespaces to origins not from packages.

**Reply by calvinrp** [reply]
Thinking through... 

If a registry wanted to fork a specific package dependency from another registry, it would remove that namespace alias so that it would resolve to packages in its registry. You'd publish (or publish copies) the packages (and package deps) that you need. 

Alternatively, the registry could change the namespace alias to another registry that has the forked packages. That might be cleaner.

**Reply by esoterra** [reply]
Ya, I think it'd be better to have a registry decide it's going to fork packages from some namespace and use one of the other techniques (e.g. log copying, log rewriting) on the whole of that namespace. It's just going to get incredibly messy if registries have a bunch of package-specific origin mappings and then you try to include that registry in another registry with package-specific mappings...

I think namespaces are a much more practical unit to map with.

**Reply by calvinrp** [reply]
I'll let others chime in. But I'm OK with that.

**Reply by lukewagner** [reply]
I think the sketch I put [above](https://github.com/bytecodealliance/registry/discussions/215#discussioncomment-7320309) ends up including this as a basic part of the design?

**Comment by calvinrp** [comment]
## Proposed API Changes:

### `Warg-Registry` HTTP Header
For all API endpoints, if the `Warg-Registry` header is present in the request, the response will be as if the registry at the domain name specified in the header value was the one responding.

The response will include the same `Warg-Registry` header and value if the server is responding as that registry. If the response header is omitted, the client should assume that the server ignored the request header.

### `GET /v1/namespace/imports`
Response is a JSON:
```json
    {
        "{namespace}": {
            "namespace": "{source-registry-namespace}",
            "registry": "{registry-domain-name}",
            "direct": true
        }
    }
```
For each imported namespace, there is an object key that is the imported `namespace` name with an object that specifies the source registry's `namespace` (omitted, if imported as the same name) and the `registry` domain name.

During package publishing, the registry will validate package imports against what is available in its own registry and these provided imported namespaces from other registries.

If the registry does not support using the `Warg-Registry` subject header for specified registry domain name, the `direct` field is set to `true`.


### `POST /v1/package/{logId}/record` (Modify Current Endpoint)
Extend the current ***publish package record*** endpoint with an optional field `namespaceImports` in the request body.

When publishing a package with an import of another registry with a namespace not currently present in the registry namespace imports, the client will include the `namespaceImports` as part of the publish request.

For each missing imported namespace, there is an object key that is the imported `namespace` name with an object that specifies the source registry's `namespace` (omitted, if imported as the same name) and the `registry` domain name.

The registry server can choose to adopt policies on whether to accept the publish or reject. If a registry does not allow importing new namespaces through this API, it will always reject publishes with unknown imports.

After publishing, the newly imported namespaces are expected to be present in the `GET /v1/namespace/imports` endpoint.

Example:
```json
{
  "packageId": "fun-corp:most-fun-pkg",
  "record": {
    "contentBytes": "ZXhhbXBsZQ==",
    "keyId": "sha256:b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c",
    "signature": "ecdsa-p256:MEUCIQCzWZBW6ux9LecP66Y+hjmLZTP/hZVz7puzlPTXcRT2wwIgQZO7nxP0nugtw18MwHZ26ROFWcJmgCtKOguK031Y1D0="
  },
  "namespaceImports": {
     "wasi": {
         "registry": "registry.wasi.dev"
     },
     "ba": {
         "registry": "registry.bytecodealliance.org",
         "namespace": "bytecode-alliance"
     }
  }
}
```
